front/lib/actions/mcp_internal_actions/servers/run_agent.ts

```diff
@@ -242,6 +248,38 @@ export default async function createServer(
             return makeMCPToolTextError(errorMessage);
           } else if (event.type === "agent_message_success") {
             break;
+          } else if (event.type === "tool_approve_execution") {
+            // We catch tool approval events and bubble them up as progress notifications to the
+            // parent tool execution.
+            // In the MCP server runner, we translate them into a tool_approve_execution event
+            // that can be ultimately shown to the end user.
+            // This part only passes along the event data without modifying them.
+            const notification: MCPProgressNotificationType = {
+              method: "notifications/progress",
+              params: {
+                progress: 0,
+                total: 1,
+                progressToken: 0,
+                data: {
+                  label: "Waiting for tool approval...",
+                  output: {
+                    type: "resource",
+                    resource: {
+                      type: "tool_approve_execution",
```

Comment:
Should we call it something more specific as it's not meant to be used for that but instead to bubble up tool approval (it's not event streamed eventually as it's swallowed)

Maybe `tool_approval_bubble_up` ?